### ISSUE TITLE
kernel BUG at fs/bcachefs/btree_iter.c:1678!

### ISSUE THREAD

**Issue by nicman23** [koverstreet/bcachefs#61]
latest master because of some issues with my 4k drives and bucket overflows

`sudo mount /dev/sda1:/dev/sdb1:/dev/nvme0n1p2 /mnt -t bcachefs -o ro,errors=continue`

[dmesg](https://github.com/koverstreet/bcachefs/files/3568196/a.txt)

please help, i have no access to my files.

show super sda1:

```
External UUID:			85bf4cd0-9945-41d8-8d72-b93c618fc93b
Internal UUID:			fafe7b2d-f6db-4b4e-823b-0ee265771dd4
Label:				
Version:			10
Created:			Tue Jan 29 14:31:26 2019
Block_size:			4.0K
Btree node size:		256.0K
Error action:			remount-ro
Clean:				0
Metadata replicas:		2
Data replicas:			2
Metadata checksum type:		crc32c (1)
Data checksum type:		crc32c (1)
Compression type:		none (0)
Foreground write target:	Group 0 (nvme)
Background write target:	Group 1 (hdd)
Promote target:			Group 0 (nvme)
String hash type:		siphash (2)
32 bit inodes:			0
GC reserve percentage:		8%
Root reserve percentage:	0%
Devices:			3 live, 3 total
Sections:			journal,members,replicas_v0,disk_groups,clean,journal_seq_blacklist
Superblock size:		11984

Members (size 176):
  Device 0:
    UUID:			3fed3b47-259c-4829-8c8c-688e751969a0
    Size:			457.3G
    Bucket size:		256.0K
    First bucket:		0
    Buckets:			1873072
    Last mount:			Tue Sep  3 07:16:36 2019
    State:			readwrite
    Group:			nvme (0)
    Data allowed:		journal,btree,data
    Has data:			(none)
    Replacement policy:		lru
    Discard:			0
  Device 1:
    UUID:			97b1b10d-f51d-4831-a007-f49fcd1794a8
    Size:			2.7T
    Bucket size:		512.0K
    First bucket:		0
    Buckets:			5723174
    Last mount:			Tue Sep  3 07:16:36 2019
    State:			readwrite
    Group:			hdd (1)
    Data allowed:		journal,btree,data
    Has data:			(none)
    Replacement policy:		lru
    Discard:			0
  Device 2:
    UUID:			da144269-9e39-49cb-a137-375c814910da
    Size:			2.7T
    Bucket size:		512.0K
    First bucket:		0
    Buckets:			5723174
    Last mount:			Tue Sep  3 07:16:36 2019
    State:			readwrite
    Group:			hdd (1)
    Data allowed:		journal,btree,data
    Has data:			(none)
    Replacement policy:		lru
    Discard:			0
```

**Comment by nicman23** [comment]
seeing a commit that reconstructing allocs need an mount option, the mount does not cause a kernel bug but it fails:
> `sudo mount /dev/sda1:/dev/sdb1:/dev/nvme0n1p2 /mnt -t bcachefs -o fix_errors,reconstruct_alloc`

[dmesg.txt](https://github.com/koverstreet/bcachefs/files/3570083/dmesg.txt)

**Comment by nicman23** [comment]
remade the array so i really cannot provide more info.

**Comment by koverstreet** [comment]
The code that manages multiple iterators needs some work; I started digging into it yesterday. Fortunately, the backtrace you provided should be all we need to diagnose this one...

**Comment by vale981** [comment]
I've had the same problem.

**Comment by hyperfekt** [comment]
Encountered on both 5a3a4087af27aa10da5f23cb174a439946153584 and 7ee280070625b0cf2528ebc461371976efa39c51, first occured after a hard reset (`kernel BUG at ../fs/bcachefs/btree_iter.c:1781`).

**Comment by nicman23** [comment]
I had a lot of hard resets due to amd GPUs being finicky but no issues with bcachefs on the 4.19 backport branch

**Comment by koverstreet** [comment]
I just pushed a rework of the relevant code to the master branch. Can someone who's been hitting it test?

**Comment by hyperfekt** [comment]
It now kernel panics instead of merely hitting a `BUG_ON`.
I'm not sure how to recover any useful info from that state. Will try and get you a dump to try yourself tomorrow.

**Comment by hyperfekt** [comment]
[Here is](https://drive.google.com/open?id=1pLckWkr1xO4MI31G9IarPnIMe6FkGkoY) [a dump](https://drive.google.com/open?id=1ZUw7OOz1aLq9cVe3bAF3cxyvAmjo5Uvq) that you should be able to use to reproduce the problem.

**Comment by koverstreet** [comment]
Were you able to get the log from the kernel panic?

**Comment by hyperfekt** [comment]
I'm not sure how to do that, but I saw it once and it was definitely in a function called something with btree iteration. Is it not reproducible with the dumps? I tested them, and it should be. I was on 2f6aaca34bb7fadeb026fcb7706b334f7259806a.

**Comment by koverstreet** [comment]
Bah, it was a bug when counting iterators for inserting extents. Fixed by 380885b0b8.
